### PR TITLE
Don't create $HOMEBREW_CACHE with sudo on Linux

### LINKED ISSUE #279: Installer leaves behind root-owned .cache folder
- [x] your problem was from running the official `install` or `uninstall` script?
- [x] after installation: ran `brew config` and `brew doctor` and included their output with your issue? If you couldn't install: provided your OS version with the output of your issue?

Output from inside the Docker container noted below:

**`brew doctor`**:
```sh
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.sh: line 4: warning: setlocale
: LC_ALL: cannot change locale (en_US.UTF-8): No such file or directory
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: "config" scripts exist outside your system or Homebrew directories.
`./configure` scripts often look for *-config scripts to determine if
software packages are installed, and which additional flags to use when
compiling and linking.

Having additional scripts in your path can confuse software installed via
Homebrew if the config script overrides a system or Homebrew-provided
script of the same name. We found the following "config" scripts:
  /bin/apt-config
/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
```

**`brew config`**:
```sh
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.sh: line 4: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8): No such file or directory
HOMEBREW_VERSION: 2.2.10
ORIGIN: https://github.com/Homebrew/brew
HEAD: 48a1637b24a0c329905a981f4cff547c7fa47943
Last commit: 2 days ago
Core tap ORIGIN: https://github.com/Homebrew/linuxbrew-core
Core tap HEAD: 6d15f07b97daad79f8140629c91d6286317ac3dc
Core tap last commit: 6 hours ago
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CACHE: /home/user1/.cache/Homebrew
HOMEBREW_LOGS: /home/user1/.cache/Homebrew/Logs
CPU: quad-core 64-bit haswell
Homebrew Ruby: 2.6.3 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3/bin/ruby
Clang: N/A
Git: 2.20.1 => /bin/git
Curl: 7.65.3 => /usr/bin/curl
Kernel: Linux 4.19.76-linuxkit x86_64 GNU/Linux
OS: Unknown
Host glibc: 2.30
/usr/bin/gcc: 9.2.1
glibc: N/A
gcc: N/A
xorg: N/A
```

## What you were trying to do (and why)
Install Homebrew on Ubuntu, and then use the `~/.config` folder as a non-root user.

## What happened (include command output)

<details>
  <summary>Command output</summary>
  <pre>

```
mkdir: cannot create directory ‘/home/user1/.cache/fisher’: Permission denied
```

  </pre>
</details>

## What you expected to happen

`~/.cache` to be owned by the user (not root) after running the install script.

## Step-by-step reproduction instructions (by running `brew` commands)
1. Delete `~/.cache`
2. Run `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"`
3. Check the permissions of `~/.cache`

Minimal Dockerfile to reproduce:
```dockerfile
FROM ubuntu:19.10

# Install brew deps and sudo
RUN apt-get update \
    && apt-get install -y \
    sudo \
    libc6 \
    gcc \
    make \
    curl \
    file \
    git

# Add a new user and add them to sudoers
# Setup sudo to be used without a password
RUN useradd --create-home user1 \
    && usermod -aG sudo user1 \
    && echo "user1 ALL=(ALL) NOPASSWD: ALL" >> /etc/sudoers

USER user1
WORKDIR /home/user1

# Commenting this out will cause a /home/user1/.cache directory to be created which is owned by root:root
# When it is not commented out, the directory is left owned by the correct user
# RUN mkdir /home/user1/.cache

RUN /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"

RUN ls -l -a /home/user1
```
Output without creating `~/.cache` first:
```sh
Step 7/7 : RUN ls -l -a /home/user1
 ---> Running in 076bd2ed92cf
total 24
drwxr-xr-x 1 user1 user1 4096 Mar 11 06:11 .
drwxr-xr-x 1 root  root  4096 Mar 11 06:11 ..
-rw-r--r-- 1 user1 user1  220 May  5  2019 .bash_logout
-rw-r--r-- 1 user1 user1 3771 May  5  2019 .bashrc
drwxr-xr-x 3 root  root  4096 Mar 11 06:11 .cache
-rw-r--r-- 1 user1 user1  807 May  5  2019 .profile
-rw-r--r-- 1 user1 user1    0 Mar 11 06:11 .sudo_as_admin_successful
```

Output with creating `~/.cache` first:
```sh
Step 8/8 : RUN ls -l -a /home/user1
 ---> Running in 346bcf4c3fcf
total 24
drwxr-xr-x 1 user1 user1 4096 Mar 11 06:14 .
drwxr-xr-x 1 root  root  4096 Mar 11 06:14 ..
-rw-r--r-- 1 user1 user1  220 May  5  2019 .bash_logout
-rw-r--r-- 1 user1 user1 3771 May  5  2019 .bashrc
drwxr-xr-x 1 user1 user1 4096 Mar 11 06:14 .cache
-rw-r--r-- 1 user1 user1  807 May  5  2019 .profile
-rw-r--r-- 1 user1 user1    0 Mar 11 06:14 .sudo_as_admin_successful
```

I think the problem is that [the line that creates `$HOMEBREW_CACHE`](https://github.com/Homebrew/install/blob/master/install.sh#L415) creates with `mkdir -p`, which creates `~/.cache` if it doesn't exist. However, [the line that runs`chown` on `$HOMEBREW_CACHE`](https://github.com/Homebrew/install/blob/master/install.sh#L421) only owns the last directory in the tree, leaving behind `~/.cache` owned by root if it didn't exist before the `mkdir` line.

### PR DESCRIPTION
**Pull request by jswny**
Don't execute `sudo` when creating `$HOMEBREW_CACHE`, so that if `~/.cache` (the parent directory of `$HOMEBREW_CACHE` on Linux) doesn't exist, the installer doesn't leave behind a root-owned `~/.cache`.

Closes #279.

Tested with the following Dockerfile:
```docker
FROM ubuntu:19.10

# Install brew deps and sudo
RUN apt-get update \
    && apt-get install -y \
    sudo \
    libc6 \
    gcc \
    make \
    curl \
    file \
    git

# Add a new user and add them to sudoers
# Setup sudo to be used without a password
RUN useradd --create-home user1 \
    && usermod -aG sudo user1 \
    && echo "user1 ALL=(ALL) NOPASSWD: ALL" >> /etc/sudoers

USER user1
WORKDIR /home/user1

# Commenting this out USED TO cause a /home/user1/.cache directory to be created which is owned by root:root
# RUN mkdir /home/user1/.cache

COPY install.sh /home/user1/install.sh
RUN /bin/bash -c "/home/user1/install.sh"

RUN ls -l -a /home/user1
```

### CODE REVIEW

File: install.sh [hunk 1]
```diff
@@ -426,7 +426,11 @@ if [[ "${#mkdirs[@]}" -gt 0 ]]; then
 fi
 
 if ! [[ -d "${HOMEBREW_CACHE}" ]]; then
-  execute_sudo "/bin/mkdir" "-p" "${HOMEBREW_CACHE}"
+  if [[ -z "${HOMEBREW_ON_LINUX-}" ]]; then
+    execute_sudo "/bin/mkdir" "-p" "${HOMEBREW_CACHE}"
+  else
+    "/bin/mkdir" "-p" "${HOMEBREW_CACHE}"
```
Comment (dawidd6) [inline install.sh]:
```suggestion
    execute "/bin/mkdir" "-p" "${HOMEBREW_CACHE}"
```

**Comment by MikeMcQuaid** [comment]
Thanks so much for your contribution! Without people like you submitting PRs we couldn't run this project. You rock, @jswny!

**Comment by jswny** [comment]
@MikeMcQuaid

Thank you for all your amazing work on Homebrew! I'm a sponsor for a reason, you and the team deserve it! I'm glad I could help!

**Comment by MikeMcQuaid** [comment]
Thank you for the kind words, @jswny!

**Pull request merged**